libs/s25main/controls/ctrlMapSelection.cpp

```diff
@@ -0,0 +1,244 @@
+// Copyright (C) 2005 - 2023 Settlers Freaks (sf-team at siedler25.org)
+//
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include "ctrlMapSelection.h"
+#include "CollisionDetection.h"
+#include "Loader.h"
+#include "RttrConfig.h"
+#include "driver/MouseCoords.h"
+#include "helpers/Range.h"
+#include "ogl/glArchivItem_Bitmap.h"
+#include <libsiedler2/ArchivItem_Bitmap.h>
+#include <libsiedler2/ColorBGRA.h>
+#include <libsiedler2/IAllocator.h>
+#include <libsiedler2/libsiedler2.h>
+
+ctrlMapSelection::ctrlMapSelection(Window* parent, unsigned id, const DrawPoint& pos, const Extent& size,
+                                   const SelectionMapInputData& inputData)
+    : Window(parent, id, pos, size), inputData(inputData), missionStatus(inputData.missionDescriptionList.size()),
+      currentSelectionPos(-1, -1), preview(false)
+{
+    missionStatus[0] = {true, false};
+    setSelectionCursor(0);
+    setupMap();
+}
+
+ctrlMapSelection::~ctrlMapSelection() = default;
+
+void ctrlMapSelection::updateEnabledMask()
+{
+    for(auto x : helpers::range(mapImages.enabledMask->getWidth()))
+    {
+        for(auto y : helpers::range(mapImages.enabledMask->getHeight()))
+        {
+            auto pixelColor = mapImages.missionMapMask->getPixel(x, y);
+            if(inputData.transparentColor == pixelColor.asValue())
+                mapImages.enabledMask->setPixel(x, y, libsiedler2::ColorBGRA(inputData.transparentColor));
+            else
+            {
+                bool playable = true;
+                for(auto idx : helpers::range(missionStatus.size()))
+                {
+                    if(!missionStatus[idx].playable
+                       && inputData.missionDescriptionList[idx].maskAreaColor == pixelColor.asValue())
+                    {
+                        playable = false;
+                        break;
+                    }
+                }
+
+                libsiedler2::ColorBGRA newPixelColor = playable ? libsiedler2::ColorBGRA(inputData.transparentColor) :
+                                                                  libsiedler2::ColorBGRA(inputData.disabledColor);
+                mapImages.enabledMask->setPixel(x, y, newPixelColor);
+            }
+        }
+    }
+}
+
+void ctrlMapSelection::setMissionsStatus(std::vector<MissionStatus> const& status)
+{
+    if(inputData.missionDescriptionList.size() != status.size())
+        throw std::runtime_error(
+          (boost::format("List has wrong size. %1% != %2%") % inputData.missionDescriptionList.size() % status.size())
+            .str());
```

Comment:
A bit easier with:

```suggestion
          helpers::format("List has wrong size. %1% != %2%", inputData.missionDescriptionList.size(), status.size()));
```

Might even safe the linebreak